plutus-contract/src/Language/Plutus/Contract/Test.hs

```diff
@@ -76,65 +92,144 @@ checkPredicate
     -> ContractTrace EmulatorAction a ()
     -> TestTree
 checkPredicate nm con predicate action =
-    HUnit.testCase nm $
+    HUnit.testCaseSteps nm $ \step ->
         case runTrace con action of
             (Left err, _) ->
                 HUnit.assertFailure $ "EmulatorAction failed. " ++ show err
-            (Right (_, st), ms) ->
-                let dt = ContractTraceResult ms st in
-                HUnit.assertBool nm (getPredicate (predicate defaultDist) dt)
+            (Right (_, st), ms) -> do
+                let dt = ContractTraceResult ms st
+                    (result, emLog) = runWriter $ unPredF predicate (defaultDist, dt)
+                if result then pure () else traverse_ step emLog
+                HUnit.assertBool nm result
 
 endpointAvailable :: Wallet -> String -> TracePredicate a
-endpointAvailable w nm _ = Predicate $ \r ->
-    nm `Set.member` Hooks.activeEndpoints (hooks w r)
+endpointAvailable w nm = PredF $ \(_, r) -> do
+    let st = Hooks.activeEndpoints (hooks w r)
+    if nm `Set.member` st
+    then pure True
+    else do
+        tellSeq ["Active endpoints:", show st,  "missing endpoint:", nm]
+        pure False
 
 interestingAddress :: Wallet -> Address -> TracePredicate a
-interestingAddress w addr _ = Predicate $ \r ->
-        addr `Set.member` Hooks.addresses (hooks w r)
-
-tx :: Wallet -> (UnbalancedTx -> Bool) -> TracePredicate a
-tx w flt _ = Predicate $ \r ->
-    any flt (Hooks.transactions (hooks w r))
-
-walletState :: Wallet -> (EM.WalletState -> Bool) -> TracePredicate a
-walletState w flt _ = Predicate $ \r ->
-    maybe False flt $ view (at w) $ EM._walletStates $  _ctrEmulatorState r
+interestingAddress w addr = PredF $ \(_, r) -> do
+    let hks = Hooks.addresses (hooks w r)
+    if addr `Set.member` hks
+    then pure True
+    else do
+        tellSeq ["Interesting addresses:", unlines (show <$> toList hks), "missing address:", show addr]
+        pure False
+
+tx :: Wallet -> (UnbalancedTx -> Bool) -> String -> TracePredicate a
+tx w flt nm = PredF $ \(_, r) -> do
+    let hks = Hooks.transactions (hooks w r)
+    if any flt hks
+    then pure True
+    else do
+        tellSeq ["Unbalanced transactions;", unlines (fmap show hks), "No transaction with '" <> nm <> "'"]
+        pure False
+
+walletState :: Wallet -> (EM.WalletState -> Bool) -> String -> TracePredicate a
+walletState w flt nm = PredF $ \(_, r) -> do
+    let ws = view (at w) $ EM._walletStates $  _ctrEmulatorState r
+    case ws of
+        Nothing -> writer (False, Seq.singleton ("Wallet state of '" <> show w <> "' not found"))
+        Just st ->
+            if flt st
+            then pure True
+            else do
+                tellSeq ["Wallet state of " <> show w <> ":", show st, "Fails '" <> nm <> "'"]
+                pure False
```

Comment:
You're a little inconsistent about when you use short `do` blocks vs `writer`. I'd probably use the `do` blocks everywhere tbh.